### PR TITLE
Support different backends on top of outlines_core

### LINKED ISSUE #1654: Integrate other structured generation backends
Outlines currently rely on `outlines-core` for creating the logits processor that does the actual logits biasing for local models. The idea would be to allow users to select alternative structured generation libraries to do that such as `xgrammar` or `llguidance`.

For this we will use a plugin system with a registry of backends.

### PR DESCRIPTION
**Pull request by RobinPicard**
Edited 2025/07/21

Closes #1654 

Not included/to be done in separate PRs:
- Use a lower level API of `xgrammar` to see whether we could have support for `LlamaCpp` and `MLXLM`
- Update the documentation on `CFG` formats (Lark vs. EBNF) and look at whether we could support format translations
- Upgrade the version of `outlines-core` used

### CODE REVIEW

File: outlines/generator.py [hunk 1]
```diff
@@ -222,16 +222,23 @@ class SteerableGenerator:
...
     """
-    logits_processor: Optional[OutlinesLogitsProcessor]
+    logits_processor: Optional[LogitsProcessorType]
 
-    def __init__(self, model: SteerableModel, output_type: Optional[Any]):
+    def __init__(
+        self,
+        model: SteerableModel,
+        output_type: Optional[Any],
+        backend_name: Optional[str] = None,
```
[the remark this answers is not in the document]
Comment (RobinPicard) [reply outlines/generator.py:231]: I wanted to call it just `backend` for the user-facing API so it's short, but then internally calling it `backend_name` to distinguish it from a `Backend` instance.
